File: app/Models/StudyGroup.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class StudyGroup extends Model
{
    use HasFactory;

    public const STATUS_ACTIVE = 'active';

    public const STATUS_ARCHIVED = 'archived';

    public const VISIBILITY_FRIENDS = 'friends';

    public const VISIBILITY_PASSWORD = 'password';

    public const VISIBILITY_PUBLIC = 'public';

    protected $fillable = [
        'owner_id',
        'name',
        'code',
        'description',
        'visibility',
        'password_hash',
        'status',
    ];

    public function getRouteKeyName(): string
    {
        return 'code';
    }

    public function owner(): BelongsTo
    {
        return $this->belongsTo(User::class, 'owner_id');
    }

    public function members(): HasMany
    {
        return $this->hasMany(StudyGroupMember::class);
    }

    public function focusRooms(): HasMany
    {
        return $this->hasMany(StudyFocusRoom::class);
    }

    public function studySessions(): HasMany
    {
        return $this->hasMany(StudySession::class);
    }

    public function isActive(): bool
    {
        return $this->status === self::STATUS_ACTIVE;
    }

    public function visibilityLabel(): string
    {
        return match ($this->visibility) {
            self::VISIBILITY_FRIENDS => 'Somente amigos',
            self::VISIBILITY_PASSWORD => 'Privado com senha',
            default => 'Publico',
        };
    }
}

File: app/models/study_group.rb
class StudyGroup < ApplicationRecord
  STATUS_ACTIVE = "active"
  STATUS_ARCHIVED = "archived"
  VISIBILITY_FRIENDS = "friends"
  VISIBILITY_PASSWORD = "password"
  VISIBILITY_PUBLIC = "public"

  belongs_to :owner, class_name: "User"
  has_many :members, class_name: "StudyGroupMember", dependent: :destroy
  has_many :focus_rooms, class_name: "StudyFocusRoom", dependent: :destroy
  has_many :study_sessions, dependent: :nullify

  validates :name, :code, :visibility, :status, presence: true
  validates :code, uniqueness: true

  scope :active, -> { where(status: STATUS_ACTIVE) }

  def to_param = code

  def active?
    status == STATUS_ACTIVE
  end

  def visibility_label
    case visibility
    when VISIBILITY_FRIENDS then "Somente amigos"
    when VISIBILITY_PASSWORD then "Privado com senha"
    else "Publico"
    end
  end

  def has_password?
    password_hash.present?
  end

  def password_matches?(password)
    return true if password_hash.blank?

    BCrypt::Password.new(password_hash) == password.to_s
  rescue BCrypt::Errors::InvalidHash
    false
  end
end
